### PR TITLE
[shopsys] Slug of Seo page is now saved as text instead of FriendlyUrl

### PR DESCRIPTION
**Pull request by TomasLudvik**
#### Description, the reason for the PR
FriendlyUrl was not best solution for saving slugs as it needs to be exact without duplicity and format controls that are used in FriendlyUrls.

#### Fixes issues
... 

#### Check the appropriate checkboxes below, please

- [ ] [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/) 
- [ ] New feature 
- [x] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

----
:globe_with_meridians: Live Preview:
  - https://tl-fix-seo-pages.odin.shopsys.cloud
  - https://cz.tl-fix-seo-pages.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Form/Admin/Seo/SeoPageFormType.php [hunk 1]
```diff
@@ -80,10 +81,17 @@ private function createBasicInformationGroup(
                 'entry_type' => TextType::class,
                 'disabled' => $seoPage !== null,
                 'required' => true,
-                'label' => t('Page URL'),
+                'label' => t('Page slug'),
                 'entry_options' => [
                     'constraints' => [
                         new Constraints\NotBlank(['message' => 'Please enter page URL']),
+                        new Constraints\Regex([
+                            'pattern' => '/^[a-z0-9\-_\/]+$/',
```
Comment (grossmannmartin) [inline packages/framework/src/Form/Admin/Seo/SeoPageFormType.php]: I think it is possible to create a page with upper-case letters

Snímek obrazovky 2024-11-21 v 10 34 33

File: packages/framework/src/Form/Constraints/UniqueSeoPageSlugValidator.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    public function validate(mixed $value, Constraint $constraint): void
+    {
+        if (!$constraint instanceof UniqueSeoPageSlug) {
+            throw new UnexpectedTypeException($constraint, UniqueCollection::class);
+        }
+
+        /** @var \Symfony\Component\Form\Form $form */
+        $form = $this->context->getObject();
+        $domainId = (int)$form->getName();
```
Comment (grossmannmartin) [inline packages/framework/src/Form/Constraints/UniqueSeoPageSlugValidator.php]: I'm not sure about this... It limits the usage of this constraint strictly to multidomain fields.
But I don't know how to do it better

File: packages/framework/src/Model/Seo/HreflangLinksFacade.php [hunk 3]
```diff
@@ -210,4 +217,20 @@ protected function getRelevantDomainIds(int $currentDomainId): array
...
+
+    /**
+     * @param int $domainId
+     * @param \Shopsys\FrameworkBundle\Model\Seo\Page\SeoPage $seoPage
+     * @return \Shopsys\FrameworkBundle\Model\Seo\HreflangLink
+     */
+    protected function createHreflangLinkForSeoPage(int $domainId, SeoPage $seoPage): HreflangLink
+    {
+        $domainRouter = $this->domainRouterFactory->getRouter($domainId);
+        $homepageUrl = $domainRouter->generate('front_homepage', [], UrlGeneratorInterface::ABSOLUTE_URL);
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Seo/HreflangLinksFacade.php]: is the homepage url the best approach? What if someone changes the homepage slug? 
isn't the domain url sufficient?

File: packages/framework/src/Model/Seo/Page/SeoPageRepository.php [hunk 4]
```diff
@@ -60,4 +62,43 @@ protected function getSeoPageRepository(): EntityRepository
...
+    /**
+     * @param int $domainId
+     * @param string $pageSlug
+     * @return \Shopsys\FrameworkBundle\Model\Seo\Page\SeoPage|null
+     */
+    public function findByDomainIdAndPageSlug(int $domainId, string $pageSlug): ?SeoPage
+    {
+        $seoPage = $this->getSeoPageRepository()
+            ->createQueryBuilder('sp')
+            ->select('sp')
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Seo/Page/SeoPageRepository.php]: this is redundant as you use repository->createQueryBuilder('sp')

File: packages/framework/src/Resources/translations/validators.cs.po [hunk 5]
```diff
@@ -362,7 +362,7 @@ msgid "Please enter password"
 msgstr "Vyplňte prosím heslo"
 
 msgid "Please enter postcode"
-msgstr ""
+msgstr "Vyplňte prosím poštovní číslo"
```
Comment (grossmannmartin) [inline packages/framework/src/Resources/translations/validators.cs.po]: usually, we use PSČ, don't we want to unify it?

File: project-base/app/src/DataFixtures/Demo/SeoPageDataFixture.php [hunk 6]
```diff
@@ -121,12 +121,12 @@ private function fillSeoPageData(
 
         $seoPageSlug = SeoPageSlugTransformer::transformFriendlyUrlToSeoPageSlug($pageSlug);
 
+        $homepageUrl = $seoPageDomainRouter->generate('front_homepage', [], UrlGeneratorInterface::ABSOLUTE_URL);
```
Comment (grossmannmartin) [inline project-base/app/src/DataFixtures/Demo/SeoPageDataFixture.php]: same for homepage here

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/framework/src/Form/Constraints/UniqueSeoPageSlugValidator.php (hunk 2 above)
Comment (TomasLudvik) [reply packages/framework/src/Form/Constraints/UniqueSeoPageSlugValidator.php]: I have reworked it

File: packages/framework/src/Model/Seo/HreflangLinksFacade.php (hunk 3 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Seo/HreflangLinksFacade.php]: Changed

File: packages/framework/src/Model/Seo/HreflangLinksFacade.php [hunk 7]
```diff
@@ -210,4 +213,17 @@ protected function getRelevantDomainIds(int $currentDomainId): array
...
+    /**
+     * @param int $domainId
+     * @param \Shopsys\FrameworkBundle\Model\Seo\Page\SeoPage $seoPage
+     * @return \Shopsys\FrameworkBundle\Model\Seo\HreflangLink
+     */
+    protected function createHreflangLinkForSeoPage(int $domainId, SeoPage $seoPage): HreflangLink
+    {
+        return new HreflangLink(
+            $this->domain->getDomainConfigById($domainId)->getLocale(),
+            $this->domain->getUrl() . $seoPage->getPageSlug($domainId),
```
Comment (grossmannmartin) [inline packages/framework/src/Model/Seo/HreflangLinksFacade.php]: I'm not really sure where I should put it, so here it is. 
It seems like we have to somehow add slashes on the proper places.

Snímek obrazovky 2024-11-25 v 17 49 23

Snímek obrazovky 2024-11-25 v 17 51 44

**Review by grossmannmartin — CHANGES_REQUESTED** [review]

File: packages/framework/src/Model/Seo/HreflangLinksFacade.php (hunk 7 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Seo/HreflangLinksFacade.php]: fixed

**Review by grossmannmartin — APPROVED** [review]

**Pull request merged**